tensorflow/core/kernels/fused_layer_norm_op.cc

```diff
@@ -0,0 +1,418 @@
+                     D_inv * x_centered)
+                        .sum(reduce_D)
+                        .eval();
+    U dmean_dx = 1.f * D_inv;
+
+    auto dx = dl_dx + dl_dvar.reshape(N_by_one).broadcast(bcast_D) * dvar_dx +
+              dl_dmean.reshape(N_by_one).broadcast(bcast_D) * dmean_dx;
+    x_backprop.device(d) = dx.template cast<T>();
+  }
+};
+
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+#define DECLARE_GPU_SPEC(T, U)                                              \
+  template <>                                                               \
+  void FusedLayerNorm<GPUDevice, T, U>::operator()(                         \
+      OpKernelContext* context, const Tensor& x_input,                      \
+      const Tensor& scale_input, const Tensor& offset_input, U epsilon,     \
+      int batch_dim, int feature_dim, Tensor* y_output,                     \
+      Tensor* saved_mean_output, Tensor* saved_inv_var_output);             \
+  extern template struct FusedLayerNorm<GPUDevice, T, U>;                   \
+  template <>                                                               \
+  void FusedLayerNormGrad<GPUDevice, T, U>::operator()(                     \
+      OpKernelContext* context, const Tensor& y_backprop_input,             \
+      const Tensor& x_input, const Tensor& scale_input,                     \
+      const Tensor& saved_mean_input, const Tensor& saved_inv_var_input,    \
+      U epsilon, int batch_dim, int feature_dim, Tensor* x_backprop_output, \
+      Tensor* scale_backprop_output, Tensor* offset_backprop_output);       \
+  extern template struct FusedLayerNormGrad<GPUDevice, T, U>;
+
+DECLARE_GPU_SPEC(float, float);
+DECLARE_GPU_SPEC(Eigen::half, float);
+
+#endif  // GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+}  // namespace functor
+
+template <typename Device, typename T, typename U>
+class FusedLayerNormOp : public OpKernel {
+ public:
+  explicit FusedLayerNormOp(OpKernelConstruction* context) : OpKernel(context) {
+    float epsilon;
+    OP_REQUIRES_OK(context, context->GetAttr("epsilon", &epsilon));
+    epsilon_ = U(epsilon);
+  }
+
+  void Compute(OpKernelContext* context) override {
+    Tensor x_input = context->input(0);
+    const Tensor& scale_input = context->input(1);
+    const Tensor& offset_input = context->input(2);
+
+    OP_REQUIRES(context, x_input.dims() == 2,
+                errors::InvalidArgument("input must be 2-dimensional",
+                                        x_input.shape().DebugString()));
+    OP_REQUIRES(context, scale_input.dims() == 1,
+                errors::InvalidArgument("scale must be 1-dimensional",
+                                        scale_input.shape().DebugString()));
+    OP_REQUIRES(context, offset_input.dims() == 1,
+                errors::InvalidArgument("offset must be 1-dimensional",
+                                        offset_input.shape().DebugString()));
+    const int batch_dim = 0;
+    const int feature_dim = 1;
```

Comment:
@kaixih Why are you restricting only 2D tensor as input? You could flatten dims [0..(n-2)] and consider as batch dimension.

Reply:
I was thinking to let the python side to reshape the tensor to 2D before calling the backend impl, since it has the `axis` value and it knows which dims should be collapsed. And I feel it might be more flexible to let the python side to handle the reshape (even transpose) for different given axis. Do you suggest we should consider the `-1` is the default feature axis and move this reshape functionality into the C++ side?

Reply:
@kaixih Yes, I prefer having reshape functionality into C++ side.